### ISSUE TITLE
Having question for the label showed by "Plotting label" in the beginning of training.

### ISSUE THREAD

**Issue by Andy718811** [ultralytics/ultralytics#13256]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am working on a project that involves using the Yolov8-segmentation model to segment the grid lines in tables. However, I’ve encountered a strange problem. I use the same method to extract labels for all tables, but some of the charts displayed in the train_batch have issues with their labels, as shown in the figure below.
![train_batch2](https://github.com/ultralytics/ultralytics/assets/87431291/62ea66f0-2885-4797-b3b7-86019d304bdc)
![train_batch1](https://github.com/ultralytics/ultralytics/assets/87431291/014cfd88-8e78-47b3-864d-cd89236ef1da)
As you can see, most of the tables are displayed normally, but for some reason, some of them appear very abnormal. The red mask indicated does not fully adhere to the white line. I don’t know what caused this. If anyone knows the possible reason, please tell me.

### Additional

######Here is my code for training, I tried a lot of combination of arguments, but still having the same problem.

```from ultralytics import YOLO
import torch
import os


os.environ['KMP_DUPLICATE_LIB_OK']='True'

torch.cuda.empty_cache()

if __name__ == '__main__':
    model = YOLO('best_yolov8s_with_holes_new_2.pt')  # load a pretrained model (recommended for training)
    # model = YOLO('./weights_laplacian_border/best_109_epoch.pt')  # load a pretrained model (recommended for training)

    results = model.train(data='config.yaml', epochs=1000, imgsz=960, batch=2,device=0,patience=150,translate=0,bgr=0,scale=0,fliplr=0,mask_ratio=2,
                          overlap_mask=False)```

###########And here is my code for retrieving the polygons from binary mask for training

```import os
import matplotlib.pyplot as plt
import matplotlib.patches as patches
from pycocotools import mask
import cv2
import numpy as np




def is_clockwise(contour):
    value = 0
    num = len(contour)
    for i, point in enumerate(contour):
        p1 = contour[i]
        if i < num - 1:
            p2 = contour[i + 1]
        else:
            p2 = contour[0]
        value += (p2[0][0] - p1[0][0]) * (p2[0][1] + p1[0][1]);
    return value < 0

def get_merge_point_idx(contour1, contour2):
    idx1 = 0
    idx2 = 0
    distance_min = -1
    for i, p1 in enumerate(contour1):
        for j, p2 in enumerate(contour2):
            distance = pow(p2[0][0] - p1[0][0], 2) + pow(p2[0][1] - p1[0][1], 2);
            if distance_min < 0:
                distance_min = distance
                idx1 = i
                idx2 = j
            elif distance < distance_min:
                distance_min = distance
                idx1 = i
                idx2 = j
    return idx1, idx2

def merge_contours(contour1, contour2, idx1, idx2):
    contour = []
    for i in list(range(0, idx1 + 1)):
        contour.append(contour1[i])
    for i in list(range(idx2, len(contour2))):
        contour.append(contour2[i])
    for i in list(range(0, idx2 + 1)):
        contour.append(contour2[i])
    for i in list(range(idx1, len(contour1))):
        contour.append(contour1[i])
    contour = np.array(contour)
    return contour

def merge_with_parent(contour_parent, contour):
    if not is_clockwise(contour_parent):
        contour_parent = contour_parent[::-1]
    if is_clockwise(contour):
        contour = contour[::-1]
    idx1, idx2 = get_merge_point_idx(contour_parent, contour)
    return merge_contours(contour_parent, contour, idx1, idx2)

def mask2polygon(image,H, W):
    contours, hierarchies = cv2.findContours(image, cv2.RETR_CCOMP, cv2.CHAIN_APPROX_TC89_KCOS)
    # lap_img=cv2.Laplacian(image, -1, 1, 1)
    # contours, hierarchies = cv2.findContours(image, cv2.RETR_CCOMP, cv2.CHAIN_APPROX_SIMPLE)
    # contours, hierarchies = cv2.findContours(image, cv2.RETR_CCOMP, cv2.CHAIN_APPROX_SIMPLE)


    # original_image = lap_img
    # contour_image = original_image.copy()
    # cv2.imshow('Contours', contour_image)
    # cv2.waitKey(0)
    contours_approx = []
    polygons = []
    for contour in contours:
        epsilon = 0.0000000000001 * cv2.arcLength(contour, True)
        contour_approx = cv2.approxPolyDP(contour, epsilon, True)
        contours_approx.append(contour_approx)

    contours_parent = []
    for i, contour in enumerate(contours_approx):
        parent_idx = hierarchies[0][i][3]
        if parent_idx < 0 and len(contour) >= 3:
            contours_parent.append(contour)
        else:
            contours_parent.append([])

    for i, contour in enumerate(contours_approx):
        parent_idx = hierarchies[0][i][3]
        if parent_idx >= 0 and len(contour) >= 3:
            contour_parent = contours_parent[parent_idx]
            if len(contour_parent) == 0:
                continue
            contours_parent[parent_idx] = merge_with_parent(contour_parent, contour)

    contours_parent_tmp = []
    for contour in contours_parent:
        if len(contour) == 0:
            continue
        
        # x=contour[0][0][0]
        # y=contour[126][0][1]
        # print('y:',y,'y stype',type(y))
        # print('len:',len(contour))
        contour = contour.astype(np.float64)
        for index in range(len(contour)):
            contour[index][0][0]=contour[index][0][0]/ W
            # print(contour[index][0][0]/ W)
            contour[index][0][1]=contour[index][0][1]/ H

        # print('contour:',contour)
        # print(type(contour))
        contours_parent_tmp.append(contour)

    polygons = []
    for contour in contours_parent_tmp:
        polygon = contour.flatten().tolist()
        polygons.append(polygon)
    return polygons 

def rle2polygon(segmentation):
    if isinstance(segmentation["counts"], list):
        segmentation = mask.frPyObjects(segmentation, *segmentation["size"])
    m = mask.decode(segmentation) 
    m[m > 0] = 255
    polygons = mask2polygon(m)
    return polygons

def remove_small_components(labels, stats, size_threshold,nlabels,mask):
    # loop over the connected components, excluding the background (label 0)
    mask=mask
    for i in range(1, nlabels):
        # get the area of the current component
        area = stats[i, cv2.CC_STAT_AREA]
        # if the area is smaller than the threshold, fill it with black
        if area < size_threshold:
            mask[labels == i] = 0
    return mask

if __name__=='__main__':
    # type = "word"
    ttype = "border"
    # input_dir = f'./datasets/data/val/pre_label/{ttype}'
    # # output_dir = f'./data/train/labels/{type}'
    # output_dir = f'./datasets/data/try/val_with_hole/labels/{ttype}'
    # input_dir = f'./datasets/data/val/pre_label/{ttype}'
    input_dir = f'./datasets/data/val/pre_label/val_border_unpadded_both_dir_and_padded'
    # D:\表格分析\yolov8_seg\datasets\data\train\pre_label\train_border_unpadded_both_dir_and_padded

    # output_dir = f'./datasets/data/val/labels_padded_640/{ttype}'
    output_dir = f'./datasets/data/val/val_labels_unpadded_both_dir_and_padded'

    for j in os.listdir(input_dir):
        image_path = os.path.join(input_dir, j)
        # load the binary mask and get its contours
        mask1 = cv2.imread(image_path, cv2.IMREAD_GRAYSCALE)
        # mask = cv2.dilate(mask, (3,3))
        # original_image = mask1
        # contour_image = original_image.copy()
        # cv2.imshow('Original', contour_image)
        # cv2.waitKey(0)

        # print(j)
        _, mask1 = cv2.threshold(mask1, 20, 255, cv2.THRESH_BINARY)
        # mask1=255-mask1

        # original_image = mask1
        # contour_image = original_image.copy()
        # cv2.imshow('Binary', contour_image)
        # cv2.waitKey(0)
        H, W = mask1.shape
        # print('mask1 shape:',mask1.shape)
        # find the connected components in the binary image
        CC = cv2.connectedComponentsWithStats(mask1, 8, cv2.CV_32S)
        
        # get the number of connected components, the labels, and the statistics
        nlabels = CC[0]
        labels = CC[1]
        stats = CC[2]
        
        # define a threshold for the size of the white regions
        size_threshold = 50
        
        # mask1=remove_small_components(labels, stats, size_threshold,nlabels,mask1)
        mask1=remove_small_components(labels, stats, size_threshold,nlabels,mask1)
        # convert the contours to polygons
        count=0
        polygons = []

        #對mask1進行膨脹
        kernel = np.ones((3,3),np.uint8)
        mask1 = cv2.dilate(mask1,kernel,iterations = 2)
        H, W = mask1.shape
        # print('mask1 shape:',mask1.shape)
        polygons=mask2polygon(mask1,H, W)
        # print(type(polygons))
        # print(polygons)

        # ''' Visualize the polygons on the original image '''
        # original_image = cv2.imread(image_path)
        # fig, ax = plt.subplots(1)
        # ax.imshow(original_image)

        # for polygon in polygons:
        #     polygon_array = [(polygon[i] * W, polygon[i + 1] * H) for i in range(0, len(polygon), 2)]
        #     polygon_patch = patches.Polygon(polygon_array, edgecolor='r', facecolor='none', linewidth=2)
        #     ax.add_patch(polygon_patch)

        # plt.show()
        ''' print the polygons '''
        if not os.path.exists(output_dir):
                os.makedirs(output_dir)
        with open('{}.txt'.format(os.path.join(output_dir, j)[:-4]), 'w') as f:
            
            for polygon in polygons:
                # print('polygon type:',type(polygon))
                # print('polygon:',(polygon))
                for p_, p in enumerate(polygon):
                    if p_ == len(polygon) - 1:
                        f.write('{}\n'.format(p))
                    elif p_ == 0:
                        # print(ttype)
                        if ttype == "word":
                            f.write('1 {} '.format(p)) 
                        # else:
                        elif ttype =="border":
                            f.write('0 {} '.format(p))    #ttype==boder(label:0)  ttype==word(label:1) 
                    else:
                        f.write('{} '.format(p))

            f.close() ```

**Comment by glenn-jocher** [comment]
@Andy718811 hello! Thanks for providing detailed information about the issue you're encountering with the label display during training using the YOLOv8-segmentation model.

From the images and description you provided, it seems like the issue might be related to how the segmentation masks are being generated or processed before being fed into the model. Here are a few things you might want to check or try:

1. **Mask Thresholding and Cleaning**: Ensure that the masks are properly thresholded and cleaned of noise before training. Small noise in the binary masks can lead to inaccurate or incomplete labels.

2. **Contour Approximation**: The method used for contour approximation can significantly affect the shape and accuracy of the predicted masks. Experiment with different epsilon values in `cv2.approxPolyDP` to find a balance between accuracy and generalization.

3. **Data Augmentation**: Check if any data augmentation techniques (like scaling, translations, or rotations) are distorting the masks in a way that affects training. Sometimes, aggressive augmentations can lead to the issues you're seeing.

4. **Label Verification**: Manually verify a subset of the labels to ensure they are correct. Errors in the automated label generation process might be propagating during training.

If you've already tried adjusting these and the problem persists, it could be beneficial to look into the model's learning rate and loss function settings to ensure they are optimized for the type of segmentation you are performing.

Keep us posted on your progress, and if you have more specific logs or error messages, feel free to share!

**Comment by Andy718811** [comment]
@glenn-jocher Thak you for your reply, I have tried the method you metioned, I double check my label with my eyes and a program, but I still can't figure it out, because the label generated looks fine to me.
Below is the code I that use to show the label
```import cv2
import numpy as np
# from google.colab.patches import cv2_imshow
import matplotlib.pyplot as plt
from PIL import Image, ImageDraw, ImageFont
from matplotlib import pyplot as plt

# ttype='word'
ttype='border'
image_path = f"./content/{ttype}/1207345172_57_form1_x1_203_y1_463_x2_2333_y2_1501_unpadded_both_dir_padded_20.png"  # path of the image, change it
annotation_path = f"./content/{ttype}/1207345172_57_form1_x1_203_y1_463_x2_2333_y2_1501_unpadded_both_dir_padded_20.txt"  # path of the annotation text file, change it

# Helper functions below assume that the image size is (640,640). Hence resizing the image.
# Open the image
img = Image.open(image_path)
# Resize the image to 640 by 640
img = img.resize((640, 640))

# <--------- Helper functions starts here ----------------------------------->

def maskVisualize(image, mask):
    fontsize = 18
    f, ax = plt.subplots(2, 1, figsize=(8, 8))
    ax[0].imshow(image)
    ax[1].imshow(mask)
    plt.show()

# Convert the mask from the txt file(annotation_path is path of txt file) to array of points making that mask.
def generate_points(annotation_path=''):
    labels = []  # this will store labels
    # assuming that the image is of dimension (640,640), then you have annotated it.
    with open(annotation_path, "r") as file:
        points = []
        for line in file:
            label, lis = line.split()[0], line.split()[1:]
            labels.append(label)
            lis = list(map(float, lis))
            for i in range(len(lis)):
                lis[i] = int(lis[i] * 640)
            newlis = []
            i = 0
            while(i < len(lis)):
                # appending the coordinates as a tuple (x,y)
                newlis.append((lis[i], lis[i+1]))
                i += 2
            points.append(newlis)
    return labels, points


# Convert the boundary coordinates to a mask array
def convert_boundary_to_mask_array(labels, points, show=0):
    # Create a new image with the same size as the desired mask
    mask = Image.new("L", (640, 640), 0)
    draw = ImageDraw.Draw(mask)
    for i, boundary_coords in enumerate(points):
        # Draw the boundary on the mask image
        draw.polygon(boundary_coords, fill=1)
        # Also put the label as text
        # Compute the centroid of the polygon
        centroid_x = sum(x for x, _ in boundary_coords) / len(boundary_coords)
        centroid_y = sum(y for _, y in boundary_coords) / len(boundary_coords)
        centroid = (int(centroid_x), int(centroid_y))
        # Write the name at the centroid
        text = str(labels[i])
        # Write the label at the centroid
        font = ImageFont.load_default()
        # text_w, text_h = draw.fontsize(text, font=font)
        # text_pos = (centroid[0] - text_w/2, centroid[1] - text_h/2)
        # draw.text(text_pos, text, font=font, fill='black')
    # Convert the mask image to a numpy array
    mask_array = np.array(mask) * 255
    # Show the mask image
    if show == 1:
        # cv2_imshow(mask_array)
        plt.imshow (mask_array)
        plt.show ()
    return mask_array


# Function that takes mask path (yolov8 seg txt file) and return mask of an image (shape of mask == shape of image)
def generate_mask(annotation_path='', show=0):
    # pass show=1 for showing the generated mask
    # firstly we generate the points (coordinates) from the annotations
    labels, points = generate_points(annotation_path)
    # once we get the points we will now generate the mask image from these points (binary mask image (black/white))
    # mask is represented by white and ground is represented as black
    mask_array = convert_boundary_to_mask_array(labels, points, show)
    return mask_array

# <---------- Helper Functions Ends here ------------------------------------------------------------->

mask_array = generate_mask(annotation_path=annotation_path, show=0)
maskVisualize(np.array(img), mask_array)```
![1207345172_57_form1_x1_203_y1_463_x2_2333_y2_1501_unpadded_both_dir_padded_20](https://github.com/ultralytics/ultralytics/assets/87431291/8a8b102b-134a-4ff5-9e05-f98715e9682e)

The label shown by the program.
![image](https://github.com/ultralytics/ultralytics/assets/87431291/5a3a1ed1-7de7-419e-beae-57b1ec6b642a)

The label shown by "Plotting label" before training start.
![train_batch1](https://github.com/ultralytics/ultralytics/assets/87431291/b4276b75-2185-48b9-946f-802f8141a0f1)

**Comment by glenn-jocher** [comment]
Hi @Andy718811, thanks for the detailed follow-up and for sharing your code. It looks like you've done a thorough job of verifying the labels visually and programmatically. If the labels appear correct in your checks but still display incorrectly in the training batches, it might be an issue related to how the data is being loaded or processed during training.

Here are a couple of additional steps you might consider:
1. **Data Loader Check**: Ensure that the data loader is correctly parsing and applying the labels and masks during training. Sometimes transformations or resizing applied during data loading can misalign the labels with the images.

2. **Debugging Output**: Temporarily modify your training script to output the images, masks, and labels just before they are fed into the model. This can help confirm whether the issue occurs during preprocessing.

3. **Version Check**: Ensure that all libraries and frameworks (especially those related to image handling like OpenCV, PIL, and PyTorch) are up to date, as sometimes bugs in these can cause unexpected behavior.

If you continue to experience issues, consider isolating a minimal example that reproduces the problem and sharing it. This might help in diagnosing the issue more effectively. Keep us posted on your progress! 🚀

**Comment by Andy718811** [comment]
@glenn-jocher Thank you for your reply. I can provide the data needed to reproduce this result, which may help us determine whether it’s an issue with the cv2 or PIL package, or a problem that occurred during preprocessing before training.

Here are the label files and original photos generated by the code I provided earlier, which are used for training. For convenience of testing, since the main focus is to see if there are any anomalies in the labels during preprocessing, I have only used this one photo for training.
![1207345172_57_form1_x1_203_y1_463_x2_2333_y2_1501_unpadded_both_dir_padded_20](https://github.com/ultralytics/ultralytics/assets/87431291/b7226249-b8be-408e-b873-d0c95e983cc9)
[1207345172_57_form1_x1_203_y1_463_x2_2333_y2_1501_unpadded_both_dir_padded_20.txt](https://github.com/user-attachments/files/15520320/1207345172_57_form1_x1_203_y1_463_x2_2333_y2_1501_unpadded_both_dir_padded_20.txt)

And here are the code for generating the label and showing the label. Please note, I changed their extensions to .txt in order to upload them.
[getting_label_from_binary_mask.txt](https://github.com/user-attachments/files/15520337/getting_label_from_binary_mask.txt)
[showing_yolov8_seg_label_mask.txt](https://github.com/user-attachments/files/15520339/showing_yolov8_seg_label_mask.txt)

Hope this could help us to find out what cause the problem.

**Comment by glenn-jocher** [comment]
Hi @Andy718811, thank you for providing the data and additional details. This will be very helpful in diagnosing the issue further. I'll review the provided label files, original photos, and your code to see if there are any anomalies during the preprocessing phase or if the issue might be related to the cv2 or PIL packages.

I'll get back to you as soon as I have an update or if I need more information. Your cooperation is greatly appreciated as we work towards resolving this! 🛠️
